packages/editor/src/lib/editor/managers/HistoryManager.ts

```diff
@@ -1,156 +1,120 @@
-	| null
-	| undefined
-	| void
-
-type ExtractData<Fn> = Fn extends CommandFn<infer Data> ? Data : never
-type ExtractArgs<Fn> = Parameters<Extract<Fn, (...args: any[]) => any>>
-
-export class HistoryManager<
-	CTX extends {
-		emit: (name: 'change-history' | 'mark-history', ...args: any) => void
-	},
-> {
-	_undos = atom<Stack<TLHistoryEntry>>('HistoryManager.undos', stack()) // Updated by each action that includes and undo
-	_redos = atom<Stack<TLHistoryEntry>>('HistoryManager.redos', stack()) // Updated when a user undoes
-	_batchDepth = 0 // A flag for whether the user is in a batch operation
-
-	constructor(
-		private readonly ctx: CTX,
-		private readonly annotateError: (error: unknown) => void
-	) {}
+import { TLHistoryBatchOptions, TLHistoryEntry } from '../types/history-types'
+import { stack } from './Stack'
 
-	onBatchComplete: () => void = () => void null
-
-	private _commands: Record<string, TLCommandHandler<any>> = {}
+enum HistoryRecorderState {
+	Recording = 'recording',
+	RecordingPreserveRedoStack = 'recordingPreserveRedoStack',
+	Paused = 'paused',
+}
 
-	getNumUndos() {
-		return this._undos.get().length
-	}
-	getNumRedos() {
-		return this._redos.get().length
-	}
-	createCommand = <Name extends string, Constructor extends CommandFn<any>>(
-		name: Name,
-		constructor: Constructor,
-		handle: TLCommandHandler<ExtractData<Constructor>>
-	) => {
-		if (this._commands[name]) {
-			throw new Error(`Duplicate command: ${name}`)
+/** @public */
+export class HistoryManager<R extends UnknownRecord> {
+	private readonly store: Store<R>
+
+	readonly dispose: () => void
+
+	private state: HistoryRecorderState = HistoryRecorderState.Recording
+	private readonly pendingDiff = new PendingDiff<R>()
+	/** @internal */
+	stacks = atom(
+		'HistoryManager.stacks',
+		{
+			undos: stack<TLHistoryEntry<R>>(),
+			redos: stack<TLHistoryEntry<R>>(),
+		},
+		{
+			isEqual: (a, b) => a.undos === b.undos && a.redos === b.redos,
 		}
-		this._commands[name] = handle
-
-		const exec = (...args: ExtractArgs<Constructor>) => {
-			if (!this._batchDepth) {
-				// If we're not batching, run again in a batch
-				this.batch(() => exec(...args))
-				return this.ctx
-			}
-
-			const result = constructor(...args)
-
-			if (!result) {
-				return this.ctx
+	)
+
+	private readonly annotateError: (error: unknown) => void
+
+	constructor(opts: { store: Store<R>; annotateError?: (error: unknown) => void }) {
+		this.store = opts.store
+		this.annotateError = opts.annotateError ?? noop
+		this.dispose = this.store.addHistoryInterceptor((entry, source) => {
+			if (source !== 'user') return
+
+			switch (this.state) {
+				case HistoryRecorderState.Recording:
+					this.pendingDiff.apply(entry.changes)
+					this.stacks.update(({ undos }) => ({ undos, redos: stack() }))
+					break
+				case HistoryRecorderState.RecordingPreserveRedoStack:
+					this.pendingDiff.apply(entry.changes)
+					break
+				case HistoryRecorderState.Paused:
+					break
```

Comment:
Hmmm yeah this is a tricky one.

I spent a while today exploring adding per-property diffs. I don't think this is workable - one scenario where that breaks is:
1. You and I are in the same room
2. I select several shapes and resize them. the top-left most shape only has its props change, so x/y stay the same. 
3. You move the shapes from one position to another
4. I undo my change

Currently (both in production and with this diff) my undoing also undoes your change - i restore my old snapshots of the shapes. With the per-property diffing, it gets weird: some shapes move back to their original positions, but others don't, or only do so on one axis:

![Kapture 2024-04-22 at 17 00 32](https://github.com/tldraw/tldraw/assets/1489520/fbdb51a7-53f8-4f82-aa28-2732bafe636c)

I don't love the idea of splitting instance state (big breaking change, also does that mean that ephemerality should be per-record instead of for certain operations?) or special undo/redo fns either... will think more about this